src/compiler/transformer.ts

```diff
@@ -0,0 +1,334 @@
+                    }
+                }
+
+                if (hoistedVariableDeclarations) {
+                    statements.push(
+                        createVariableStatement(
+                            createVariableDeclarationList(hoistedVariableDeclarations)
+                        )
+                    );
+                }
+            }
+
+            // Restore the previous lexical environment.
+            lexicalEnvironmentStackOffset--;
+            hoistedVariableDeclarations = lexicalEnvironmentVariableDeclarationsStack[lexicalEnvironmentStackOffset];
+            hoistedFunctionDeclarations = lexicalEnvironmentFunctionDeclarationsStack[lexicalEnvironmentStackOffset];
+            return statements;
+        }
+    }
+
+    /**
+     * High-order function, creates a function that executes a function composition.
+     * For example, `chain(a, b)` is the equivalent of `x => ((a', b') => y => b'(a'(y)))(a(x), b(x))`
+     *
+     * @param args The functions to chain.
+     */
+    function chain<T, U>(...args: ((t: T) => (u: U) => U)[]): (t: T) => (u: U) => U;
+    function chain<T, U>(a: (t: T) => (u: U) => U, b: (t: T) => (u: U) => U, c: (t: T) => (u: U) => U, d: (t: T) => (u: U) => U, e: (t: T) => (u: U) => U): (t: T) => (u: U) => U {
+        if (e) {
+            const args = arrayOf<(t: T) => (u: U) => U>(arguments);
+            return t => compose(...map(args, f => f(t)));
+        }
+        else if (d) {
+            return t => compose(a(t), b(t), c(t), d(t));
+        }
+        else if (c) {
+            return t => compose(a(t), b(t), c(t));
+        }
+        else if (b) {
+            return t => compose(a(t), b(t));
+        }
+        else if (a) {
+            return t => compose(a(t));
+        }
+        else {
+            return t => u => u;
+        }
+    }
+
+    /**
+     * High-order function, composes functions. Note that functions are composed inside-out;
+     * for example, `compose(a, b)` is the equivalent of `x => b(a(x))`.
+     *
+     * @param args The functions to compose.
+     */
+    function compose<T>(...args: ((t: T) => T)[]): (t: T) => T;
+    function compose<T>(a: (t: T) => T, b: (t: T) => T, c: (t: T) => T, d: (t: T) => T, e: (t: T) => T): (t: T) => T {
+        if (e) {
+            const args = arrayOf(arguments);
+            return t => reduceLeft<(t: T) => T, T>(args, (u, f) => f(u), t);
+        }
+        else if (d) {
+            return t => d(c(b(a(t))));
+        }
+        else if (c) {
+            return t => c(b(a(t)));
+        }
+        else if (b) {
+            return t => b(a(t));
+        }
+        else if (a) {
+            return t => a(t);
+        }
+        else {
+            return t => t;
+        }
+    }
+
+    /**
+     * Makes an array from an ArrayLike.
+     */
+    function arrayOf<T>(arrayLike: ArrayLike<T>) {
+        const array: T[] = [];
+        for (let i = 0; i < arrayLike.length; i++) {

```

Comment:
Why not just
```ts
function arrayOf<T>(arrayLike: ArrayLike<T>) {
	return Array.prototype.slice.call(arrayLike);	
}
```

I thought that was fairly canonical as the way to make arraylike things into arrays.